Option - ask for confirmation on branch checkout

IlijaQ opened the pull request:
Fixes #11116

## Proposed changes

- In Settings / Advanced / Confirmations / section Branches added `Checkout branch using left panel` checkbox (disabled by default). 
- By checking it, every time a user double-clicks on a branch on left panel, a message box will prompt asking to confirm if the user wants to checkout branch.

## Screenshots

![ConfirmationSettingsForm](https://github.com/gitextensions/gitextensions/assets/79657913/664207c2-f830-4601-9923-5cb983894ba4)
![BranchCheckoutConfirmationMessageBox5](https://github.com/gitextensions/gitextensions/assets/79657913/0e6841a4-3be3-4bcd-9356-6c808a9c8a72)


## Test methodology

- Manual with GE repo
- Tried closing and reopening the application. The selected option is saved after closing and reopening.

## Test environment(s)

- Git Extensions 33.33.33
- Build 6a64d95
- Git 2.30.0.windows.2 (recommended: 2.41.0 or later)
- Microsoft Windows NT 10.0.19045.0
- .NET 6.0.14
- DPI 96dpi (no scaling)

## Merge strategy

I agree that the maintainer squash merge this PR (if the commit message is clear).

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

IlijaQ commented:
Hi Team,

I have few issues I want to discuss.

I have that issue while adding checkbox via graphical interface, VS automatically changes the ConfirmationSettingsPage.Designer.cs mostly erasing `this` keyword from all fields in file. I have tried to counteract it by reversing changes in formCommit before committing the file, but some changes in the designer I had to make by hand e.g. in lines 145 to 204 where new row is added for the new checkbox - reversing the changes will remove the room made for checkbox previously. I know that it is not recommended to manually edit the designer. What do you suggest on this?

Since we are adding an option to settings, will this reset the user settings for users updating from older versions? I suppose no, but just in case, I need to ask?

I have used TaskDialogIcon.Warning as a MessageBox icon, Is TaskDialogIcon.Information a more suitable icon?

gerhardol commented:
Put this with the other confirmations
Add a possibility to deactivate from the dialog if enabled by default, see some other of these popups

![image](https://github.com/gitextensions/gitextensions/assets/6248932/64b271d7-ecb4-45d3-8fdb-665998447a5e)

mstv commented:
> VS automatically changes the ConfirmationSettingsPage.Designer.cs mostly erasing this keyword from all fields in file.

I use to save the unmodified dialog and commit it separately first.

IlijaQ commented:
> Put this with the other confirmations

I didn’t understand, I put it under confirmations section. Can you, please, clarify?
![ConfirmationSettingsForm2](https://github.com/gitextensions/gitextensions/assets/79657913/126945fb-58bc-46da-b8f4-7ff984e21ee4)

> Add a possibility to deactivate from the dialog if enabled by default, see some other of these popups

I didn’t include a possibility to deactivate from the dialogue because I made this option disabled by default. If you think that this option should be enabled by default or that the possibility to deactivate from the dialogue should be there anyway, I’ll make the necessary changes.

gerhardol commented:
> I didn’t understand, I put it under confirmations section. Can you, please, clarify? ![ConfirmationSettingsForm2](https://user-images.githubusercontent.com/79657913/264778222-126945fb-58bc-46da-b8f4-7ff984e21ee4.png)

> I didn’t include a possibility to deactivate from the dialogue because I made this option disabled by default. If you think that this option should be enabled by default or that the possibility to deactivate from the dialogue should be there anyway, I’ll make the necessary changes.

Fine I saw some other position for the setting. It was probably too late.

For VS changes, you can make an initial commit with the automatic changes if transferring wanted changes is too much.

Code looks reasonable.

RussKie commented on GitCommands/Settings/AppSettings.cs:
```diff
@@ -1839,6 +1839,12 @@ public static DiffListSortType DiffListSorting
             set => SetEnum("DiffListSortType", value);
         }
 
+        public static bool DontConfirmBranchCheckout
+        {
+            get => GetBool("DontConfirmBranchCheckout", true);
+            set => SetBool("DontConfirmBranchCheckout", value);
+        }
```
I _think_ all other similar settings are grouped together. If so, please put the new settings near those.

mstv replied:
For new settings, please use:
```suggestion
        public static readonly ISetting<bool> DontConfirmBranchCheckout= Setting.Create(ConfirmationsSettingsPath , nameof(DontConfirmBranchCheckout), true);
```

Needs to be added to the top of the file:
```
        private static readonly SettingsPath ConfirmationsSettingsPath = new AppSettingsPath("Confirmations");
```

IlijaQ replied:
Some of the settings are grouped together, but not all of them. After applying the suggested code changes, I put the new setting near some of the branch settings.

RussKie commented on GitUI/MessageBoxes.cs:
```diff
@@ -127,6 +131,21 @@ public static bool ConfirmUpdateSubmodules(IWin32Window? owner)
+                Text = Instance._confirmBranchCheckout.Text,
+                Heading = Instance._confirmCheckoutHeader.Text,
+                Caption = Instance._confirmCheckout.Text,
+                Icon = TaskDialogIcon.Warning,
```
Why is it a warnings? We use the info icon in other dialogs.

mstv replied:
Please use `MessageBoxes.Confirm()`.

Feel free to make `MessageBoxes.Show()` use a `TaskDialog` instead of a `MessageBox` - rather in a separate PR though.

mstv commented on GitUI/CommandsDialogs/SettingsDialog/Pages/ConfirmationsSettingsPage.cs:
```diff
@@ -24,6 +24,7 @@ protected override void SettingsToPage()
             chkPushNewBranch.Checked = !AppSettings.DontConfirmPushNewBranch;
             chkAddTrackingRef.Checked = !AppSettings.DontConfirmAddTrackingRef;
             chkBranchDeleteUnmerged.Checked = !AppSettings.DontConfirmDeleteUnmergedBranch;
+            chkBranchCheckoutConformation.CheckState = ToCheckboxStateInverted(AppSettings.DontConfirmBranchCheckout);
```
```suggestion
            chkBranchCheckoutConformation.CheckState = ToCheckboxStateInverted(AppSettings.DontConfirmBranchCheckout.Value);
```

IlijaQ commented:
Ok, here's what I have changed:

I have reverted my changes to the Designer since it was changed by hand. I added an initial commit with the automatic changes and another commit with changes made in graphical design, as @mstv and @gerhardol suggested.

Applyed code changes to `AppSettings.cs` and `ConfirmationSettingsPage.cs` as @mstv suggested.

> I think all other similar settings are grouped together. If so, please put the new settings near those.

Some of the settings are grouped together, but not all of them. After applying the suggested code changes, I put the new setting near some of the branch settings.

Message box implemented with `MessageBoxes.Confirm()` according to your suggestion.
![BranchCheckoutConfirmationMessageBox3](https://github.com/gitextensions/gitextensions/assets/79657913/135b9f4c-4150-4e76-a0be-24a020eada7b)
The screenshot in the PR description is updated to represent the new message box.

Text added to English.xlf

mstv commented on GitUI/MessageBoxes.cs:
```diff
@@ -46,6 +46,9 @@ public class MessageBoxes : Translate
         private readonly TranslationString _cannotOpenGitExtensionsCaption = new("Cannot open Git Extensions");
 
+        private readonly TranslationString _confirmBranchCheckoutCaption = new("Confirm checkout");
+        private readonly TranslationString _confirmBranchCheckout = new("Are you sure you want to check out this branch?");
```
❔ name the branch in the confirmation
```suggestion
        private readonly TranslationString _confirmBranchCheckout = new(@"Are you sure you want to check out branch ""{0}""?");
```

mstv commented on GitUI/MessageBoxes.cs:
```diff
@@ -127,6 +130,11 @@ public static bool ConfirmUpdateSubmodules(IWin32Window? owner)
         }
 
+        public static bool ConfirmBranchCheckout(IWin32Window? owner)
+        {
+            return Confirm(owner, Instance._confirmBranchCheckout.Text, Instance._confirmBranchCheckoutCaption.Text);
+        }
```
```suggestion
        => Confirm(owner, string.Format(Instance._confirmBranchCheckout.Text, branchName), Instance._confirmBranchCheckoutCaption.Text);
```
with `branchName` as additional argument

mstv commented on GitUI/Translation/English.xlf:
```diff
@@ -1053,6 +1053,10 @@ Please make sure that Git for Windows is installed or set the correct command ma
         <target />
       </trans-unit>
+      <trans-unit id="chkBranchCheckoutConformation.Text">
+        <source>Always ask for confirmation on branch checkout</source>
```
```suggestion
        <source>Checkout branch using left panel</source>
```

gerhardol commented on GitUI/CommandsDialogs/SettingsDialog/Pages/ConfirmationsSettingsPage.Designer.cs:
```diff
@@ -28,346 +28,383 @@ protected override void Dispose(bool disposing)
+            chkBranchCheckoutConformation.Location = new Point(3, 245);
+            chkBranchCheckoutConformation.Name = "chkBranchCheckoutConformation";
+            chkBranchCheckoutConformation.Size = new Size(282, 19);
+            chkBranchCheckoutConformation.TabIndex = 22;
```
TabIndex was added, this is now incorrect.
Also move to the position in the form

gerhardol reviewed (commented):
A few comments, otherwise fine to me.
As this is a draft, I suggest you fix the comments, squash and rebase on master before removing draft.

mstv commented on GitUI/LeftPanel/RepoObjectsTree.cs:
```diff
@@ -632,6 +632,11 @@ private void SelectNode(NodeBase node, bool multiple, bool includingDescendants)
 
         private void OnNodeDoubleClick(object sender, TreeNodeMouseClickEventArgs e)
         {
+            if (!AppSettings.DontConfirmBranchCheckout.Value && !MessageBoxes.ConfirmBranchCheckout(this))
+            {
+                return;
+            }
+
```
This is too early here - definitely not to be done before HitTest.
And it shall affect branches only.
Hence it needs to be moved to both `LocalBranchNode.Checkout` and `RemoteBranchNode.Checkout`.
The check for `AppSettings.DontConfirmBranchCheckout.Value` can be moved to `MessageBoxes.ConfirmBranchCheckout`.

mstv reviewed (commented):
I have noticed a few more points.

gerhardol commented on GitUI/CommandsDialogs/SettingsDialog/Pages/ConfirmationsSettingsPage.cs:
```diff
@@ -24,6 +24,7 @@ protected override void SettingsToPage()
             chkPushNewBranch.Checked = !AppSettings.DontConfirmPushNewBranch;
             chkAddTrackingRef.Checked = !AppSettings.DontConfirmAddTrackingRef;
             chkBranchDeleteUnmerged.Checked = !AppSettings.DontConfirmDeleteUnmergedBranch;
+            chkBranchCheckoutConformation.CheckState = ToCheckboxStateInverted(AppSettings.DontConfirmBranchCheckout.Value);
```
I think this was intentional, this is a tri-state box

mstv replied:
This is a global settings, i.e. without other sources nor inheritance. Unset `bool` settings make use of the default value given to `Setting.Create`.
(All those other confirmation checkboxes around do not use `ToCheckboxStateInverted`.)

IlijaQ commented on GitCommands/Settings/AppSettings.cs:
```diff
@@ -1003,6 +1004,8 @@ public static bool DontConfirmCommitIfNoBranch
             set => SetBool("DontConfirmCommitIfNoBranch", value);
         }
 
+        public static readonly ISetting<bool> DontConfirmBranchCheckout = Setting.Create(ConfirmationsSettingsPath, nameof(DontConfirmBranchCheckout), true);
```
> This will make this setting separate from similar settings though

Yes, almost all settings in AppSettings.cs are named as negative, but ok, I’ll make the changes as you suggested.

mstv replied:
Yes, the old settings names have been kept for downward compatibility with existing setting files of the users.

IlijaQ commented:
OK, thanks for the detailed feedback. I’ll update the code as you suggested.

pmiossec commented:
@IlijaQ Seems good except, please run `update-loc.cmd`. It seems that the translation file is not up to date... (CI is not happy.)
You could also maybe remove the draft state of this PR.

Side note: it seems that the last version of Visual Studio remove the `this.` on the generated `.designer.cs` file making even more a pain to read the diff. Maybe we should do a PR to remove it to all files. What do you think? I maybe could do that if everyone agrees...

IlijaQ commented:
I ran `update-loc.cmd` it appears to have worked. Thank you.

Removed the draft state of PR.

> Maybe we should do a PR to remove it to all files. What do you think?

I don’t know if the majority of currently active contributors are using the latest version of VS that does this. If so, I think it would be fine.

mstv commented on GitCommands/Settings/AppSettings.cs:
```diff
@@ -33,6 +33,7 @@ public static class AppSettings
         public static DistributedSettings SettingsContainer { get; private set; }
 
         private static readonly SettingsPath DetailedSettingsPath = new AppSettingsPath("Detailed");
+        private static readonly SettingsPath ConfirmationsSettingsPath = new AppSettingsPath("Confirmations");
```
Class members of the same category should be sorted alphabetically (when added - although there are lots of deviations from this style rule in the existing code).
```suggestion
        private static readonly SettingsPath ConfirmationsSettingsPath = new AppSettingsPath("Confirmations");
        private static readonly SettingsPath DetailedSettingsPath = new AppSettingsPath("Detailed");
```

mstv commented on GitUI/LeftPanel/LocalBranchNode.cs:
```diff
@@ -54,6 +54,11 @@ internal override void OnDelete()
 
         public bool Checkout()
         {
+            if (!MessageBoxes.ConfirmBranchCheckout(null, FullPath))
+            {
+                return false;
+            }
+
             return UICommands.StartCheckoutBranch(ParentWindow(), branch: FullPath, remote: false);
```
```suggestion
            return MessageBoxes.ConfirmBranchCheckout(ParentWindow(), FullPath)
                && UICommands.StartCheckoutBranch(ParentWindow(), branch: FullPath, remote: false);
```

mstv commented on GitUI/LeftPanel/RemoteBranchNode.cs:
```diff
@@ -63,6 +63,11 @@ public bool Delete()
 
         public bool Checkout()
         {
+            if (!MessageBoxes.ConfirmBranchCheckout(null, FullPath))
+            {
+                return false;
+            }
+
             return UICommands.StartCheckoutRemoteBranch(TreeViewNode.TreeView, FullPath);
```
ditto

mstv commented on GitUI/MessageBoxes.cs:
```diff
@@ -127,6 +130,16 @@ public static bool ConfirmUpdateSubmodules(IWin32Window? owner)
         }
 
+        public static bool ConfirmBranchCheckout(IWin32Window? owner, string branchName)
+        {
+            if (AppSettings.ConfirmBranchCheckout.Value)
+            {
+                return Confirm(owner, string.Format(Instance._confirmBranchCheckout.Text, branchName), Instance._confirmBranchCheckoutCaption.Text);
+            }
+
+            return true;
+        }
```
```suggestion
        => AppSettings.ConfirmBranchCheckout.Value
            ? Confirm(owner, string.Format(Instance._confirmBranchCheckout.Text, branchName), Instance._confirmBranchCheckoutCaption.Text)
            : true;
```

mstv reviewed (commented):
final nits

gerhardol reviewed (commented):
+2 after mstv changes

RussKie commented on GitUI/LeftPanel/LocalBranchNode.cs:
```diff
@@ -54,6 +54,11 @@ internal override void OnDelete()
 
         public bool Checkout()
         {
+            if (!MessageBoxes.ConfirmBranchCheckout(null, FullPath))
```
❗  Messages must always have a parent

RussKie reviewed (changes requested):
As per @mstv reviews

RussKie commented:
> Side note: it seems that the last version of Visual Studio remove the `this.` on the generated `.designer.cs` file making even more a pain to read the diff. Maybe we should do a PR to remove it to all files. What do you think? I maybe could do that if everyone agrees...

That'd be great, thank you.

/cc: @merriemcgaw FYI

mstv commented on GitUI/LeftPanel/LocalBranchNode.cs:
```diff
@@ -1,4 +1,5 @@
 ﻿using System.Diagnostics;
+using GitUI.CommandsDialogs;
```
Should be removed
```suggestion
```

mstv commented on GitUI/MessageBoxes.cs:
```diff
@@ -127,6 +130,10 @@ public static bool ConfirmUpdateSubmodules(IWin32Window? owner)
         }
 
+        public static bool ConfirmBranchCheckout(IWin32Window? owner, string branchName)
+            => AppSettings.ConfirmBranchCheckout.Value
+            ? Confirm(owner, string.Format(Instance._confirmBranchCheckout.Text, branchName), Instance._confirmBranchCheckoutCaption.Text) : true;
```
Support Left Reading, please!
The long one-liners in the `Checkout` methods are at my limit.
But this one should match our code style:
```suggestion
            => AppSettings.ConfirmBranchCheckout.Value
                ? Confirm(owner, string.Format(Instance._confirmBranchCheckout.Text, branchName), Instance._confirmBranchCheckoutCaption.Text)
                : true;
```
or a little mind-bending:
```suggestion
            => !AppSettings.ConfirmBranchCheckout.Value || Confirm(owner, string.Format(Instance._confirmBranchCheckout.Text, branchName), Instance._confirmBranchCheckoutCaption.Text);
```

IlijaQ replied:
Maybe we should break the parameters of the `Confirm` method into separate lines as done by calling `Show` method in the `ShowGitConfigurationExceptionMessage` method in the same file. What do you think?

```suggestion
        public static bool ConfirmBranchCheckout(IWin32Window? owner, string branchName)
            => AppSettings.ConfirmBranchCheckout.Value
                ? Confirm(owner,
                    string.Format(Instance._confirmBranchCheckout.Text, branchName),
                    Instance._confirmBranchCheckoutCaption.Text)
                : true;
```

mstv replied:
In principle, yes. 😀 This `string.Format` is of low interest though. And we prefer to use the available screen width for arguments then.

RussKie commented:
> Side note: it seems that the last version of Visual Studio remove the `this.` on the generated `.designer.cs` file making even more a pain to read the diff. Maybe we should do a PR to remove it to all files. What do you think? I maybe could do that if everyone agrees...

The designer is now ".editorconfig"-aware, meaning if `this.` is discouraged via the .editorconfig, then the designer will remove `this.` from the designer files.
Perhaps, we could try something like that and see if that'd work:
```
[*.Designer.cs]
dotnet_style_qualification_for_field = true:suggestion
```

RussKie commented:
Thank you
